pkg/apis/clusterregistry/v1alpha1/types.go

```diff
@@ -167,3 +171,87 @@ type ClusterCondition struct {
 	// +optional
 	Message string `json:"message,omitempty" protobuf:"bytes,6,opt,name=message"`
 }
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterCredentials combines a Cluster record with a set of credentials
+// (stored in a kubeconfig) that can be used to access the cluster. The status
+// of ClusterCredentials reflects whether those credentials can be used - from
+// within the cluster hosting the ClusterCredentials resource - to access the
+// cluster's healthz endpoint.
+//
+// +k8s:openapi-gen=true
+// +kubebuilder:resource:path=clustercredentials
+// +kubebuilder:subresource:status
+type ClusterCredentials struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterCredentialsSpec   `json:"spec,omitempty"`
+	Status ClusterCredentialsStatus `json:"status,omitempty"`
+}
+
+// ClusterCredentialsSpec defines the desired state of a ClusterCredentials.
+type ClusterCredentialsSpec struct {
+	// ClusterRef is a reference to the Cluster to access.
+	ClusterRef v1.LocalObjectReference `json:"clusterRef,omitempty"`
+
+	// SecretRef is a reference to the secret in the namespace that contains a
+	// kubeconfig with credentials access the referenced cluster. The secret
+	// must have a key named "kubeconfig" that contains a valid kubeconfig.
```

Comment:
It seems strange to me to have an entire kubeconfig file in an object that is called "Credentials". A kubeconfig file includes:

* `clusters` - an array of connection information (endpoints, TLS details)
* `users` - an array of credentials used for AuthN (tokens, client certs, instructions to execute CLI binaries, etc.)
* `contexts` - a combination of `cluster` + `user`, optionally with a namespace
* `current-context` - (optional) currently selected context

Of those, only entries in the `users` array make sense from the perspective of Credentials.

If the kubeconfig file has one or more `clusters` defined in it, should I client tool use one of those endpoints, or the endpoints specified in the `ClusterSpec.KubernetesAPIEndpoints` object?

I think it would make more sense to have a secret that contained a single `user` object in the kubeconfig file. Then it would be possible to construct a kubeconfig file using a combination of the `KubernetesAPIEndpoints` objects and `ClusterCredentialsSpec` objects.

Reply:
That's a good suggestion. I'll incorporate this into the doc for this field (I don't believe it should change the API as currently written from a structural standpoint).